### PR TITLE
Define arithmetic and bitwise operations

### LINKED ISSUE #257: Builtin operators
Some data formats require integer operations to be performed during parsing. It would be nice to implement some of these in Fathom!

Arithmetic operations:

- `global add_int : Int -> Int -> Int`
- `global sub_int : Int -> Int -> Int`
- `global mul_int : Int -> Int -> Int`
- `global div_int : Int -> Int -> Int`
- `global mod_int : Int -> Int -> Int`
- `global neg_int : Int -> Int`

Relational operations:

- `global lt_int : Int -> Int -> Bool`
- `global le_int : Int -> Int -> Bool`
- `global gt_int : Int -> Int -> Bool`
- `global ge_int : Int -> Int -> Bool`
- `global eq_int : Int -> Int -> Bool`
- `global ne_int : Int -> Int -> Bool`

Bitwise operations:

- `global shl_int : Int -> Int -> Int`
- `global shr_int : Int -> Int -> Int`
- `global and_int : Int -> Int -> Int`
- `global or_int : Int -> Int -> Int`
- `global xor_int : Int -> Int -> Int`

Logical operations:

- `global not_bool : Bool -> Bool`
- `global and_bool : Bool -> Bool -> Bool`
- `global and_bool : Bool -> Bool -> Bool`

Implementing some operators in the surface language might make these look a bit nicer for users!

Eventually it would be nice to give these precise types using [refinement types](https://github.com/yeslogic/fathom/issues/258), which would let us avoid overflows and wrapping issues in parsers (a common source of security vulnerabilities). It might be worth looking at the implementations of other languages like [SPRITE](https://github.com/ranjitjhala/sprite-lang), [Liquid Haskell](https://github.com/FStarLang/FStar/), [F*](https://github.com/FStarLang/FStar/), [Wuffs](https://github.com/google/wuffs/), [Dafny](https://github.com/dafny-lang/dafny), [ATS](https://github.com/githwxi/ATS-Postiats), etc. to get an idea for how we might want to do this.

### PR DESCRIPTION
**Pull request by brendanzab**
This introduces a bunch of primitive operations for signed and unsigned integers. I need to figure out a better way of managing these primitives (I have some ideas about this), but I'm working with what I have right now!

Todo:

- [x] primitive variants
- [x] add to elaboration environment
- [x] evaluation steps for primitives

Future work:

- Consider what to do about underflow, overflow, and division by zero errors (see: #258)
- Add infix operators for addition, subtraction, multiplication, and division

Closes #257

### CODE REVIEW

**Comment by mikeday** [comment]
Do we need bitwise ops for signed integer types or it would it be better to omit those and add casts to unsigned types later?

**Comment by brendanzab** [comment]
Ahh yes, that'd make more sense. Thanks!

File: fathom/src/core/semantics.rs [hunk 1]
```diff
@@ -330,6 +330,107 @@ impl<'arena, 'env> EvalContext<'arena, 'env> {
...
+    match prim {
+        Prim::FormatRepr => step!(context, [format] => context.apply_repr(format)),
+
+        Prim::U8Add => const_step!([x: U8, y: U8] => Const::U8(u8::checked_add(*x, *y)?)),
+        Prim::U8Sub => const_step!([x: U8, y: U8] => Const::U8(u8::checked_sub(*x, *y)?)),
+        Prim::U8Mul => const_step!([x: U8, y: U8] => Const::U8(u8::checked_mul(*x, *y)?)),
+        Prim::U8Div => const_step!([x: U8, y: U8] => Const::U8(u8::checked_div(*x, *y)?)),
+        Prim::U8Not => const_step!([x: U8] => Const::U8(u8::not(*x))),
+        Prim::U8Shl => const_step!([x: U8, y: U8] => Const::U8(u8::shl(*x, *y))), // TODO: Use `checked_shl`
+        Prim::U8Shr => const_step!([x: U8, y: U8] => Const::U8(u8::shr(*x, *y))), // TODO: Use `checked_shr`
```
Comment (brendanzab) [inline fathom/src/core/semantics.rs:368]: Need to decide on the types I expect here. Rust seems to expect `u32`s on the rhs of [`checked_shl`](https://doc.rust-lang.org/std/primitive.u8.html#method.checked_shl) and [`checked_shr`](https://doc.rust-lang.org/std/primitive.u8.html#method.checked_shr)?

Comment (wezm) [reply fathom/src/core/semantics.rs:368]: `u32` allows for a lot of shifts!

Comment (brendanzab) [reply fathom/src/core/semantics.rs:368]: I know! :O

Comment (brendanzab) [reply fathom/src/core/semantics.rs:368]: is `U8` ok? Surely? 🤔

Comment (wezm) [reply fathom/src/core/semantics.rs:368]: "`U8` ought to be enough for anybody"

Comment (brendanzab) [reply fathom/src/core/semantics.rs:368]: Might just do some conversions

**Comment by brendanzab** [comment]
Ok, this should be good to go!

**Review by mikeday — APPROVED** [review]
Looks good!

**Review by wezm — COMMENTED** [review]
Wow, you got these done super quick, nice 👍 Looks good to my untrained eyes.

**Pull request merged**